src/waveform/renderers/waveformmark.cpp

```diff
@@ -127,14 +136,242 @@ bool WaveformMark::contains(QPoint point, Qt::Orientation orientation) const {
+                m_contentRect = metrics.tightBoundingRect(label);
+
+                // Now we calculate how much bigger this is than our target
+                // size.
+                const auto ratioH = targetHeight / m_contentRect.height();
+                const auto ratioW = targetWidth / m_contentRect.width();
+                const auto ratio = std::min(ratioH, ratioW);
+
+                // And we scale the font size accordingly.
+                m_font.setPointSizeF(50.0 * ratio);
+                metrics = QFontMetricsF(m_font);
+                m_contentRect = metrics.tightBoundingRect(label);
+            } else if (useIcon) {
+                m_contentRect = QRectF(0.f, 0.f, std::ceil(capHeight), std::ceil(capHeight));
+            } else {
+                m_contentRect = QRectF{0.f,
+                        -capHeight,
+                        metrics.boundingRect(label).width(),
+                        capHeight};
+            }
+        }
+
+        const Qt::Alignment alignH = align & Qt::AlignHorizontal_Mask;
+        const Qt::Alignment alignV = align & Qt::AlignVertical_Mask;
+        const bool alignHCenter{alignH == Qt::AlignHCenter};
+        const qreal widthRounding{alignHCenter ? 2.f : 1.f};
+
+        m_labelRect = QRectF{0.f,
+                0.f,
+                std::ceil((m_contentRect.width() + 2.f * margin) / widthRounding) *
+                        widthRounding,
+                std::ceil(capHeight + 2.f * margin)};
+
+        m_imageSize = QSizeF{alignHCenter ? m_labelRect.width() + 1.f
+                                          : 2.f * m_labelRect.width() + 1.f,
+                breadth};
+
+        if (alignH == Qt::AlignHCenter) {
+            m_labelRect.moveLeft((m_imageSize.width() - m_labelRect.width()) / 2.f);
+        } else if (alignH == Qt::AlignRight) {
+            m_labelRect.moveRight(m_imageSize.width() - 0.5f);
+        } else {
+            m_labelRect.moveLeft(0.5f);
+        }
+
+        if (alignV == Qt::AlignVCenter) {
+            m_labelRect.moveTop((m_imageSize.height() - m_labelRect.height()) / 2.f);
+        } else if (alignV == Qt::AlignBottom) {
+            m_labelRect.moveBottom(m_imageSize.height() - 0.5f);
+        } else {
+            m_labelRect.moveTop(0.5f);
+        }
+    }
+    QSize getImageSize(float devicePixelRatio) const {
+        return QSize{static_cast<int>(m_imageSize.width() * devicePixelRatio),
+                static_cast<int>(m_imageSize.height() * devicePixelRatio)};
+    }
+};
+
+QImage WaveformMark::generateImage(float breadth, float devicePixelRatio) {
+    // Load the pixmap from file.
+    // If that succeeds loading the text and stroke is skipped.
+
+    m_breadth = static_cast<int>(breadth);
+
+    if (!m_pixmapPath.isEmpty()) {
+        QString path = m_pixmapPath;
+        // Use devicePixelRatio to properly scale the image
+        QImage image = *WImageStore::getImage(path, devicePixelRatio);
+        // QImage image = QImage(path);
```

Comment:
please remove commented-out lines like this throughout

Reply:
Done